### PR TITLE
✨(flower) create a flower application to monitor celery

### PR DESCRIPTION
**Pull request by lunika**
## Purpose

We want to monitor celery workers used by edxapp. Flower is tool written to
do this job. We created a non blue-green app to add them in a stack
alongside edxapp

## Proposal

- [x] create flower application

### CODE REVIEW

File: apps/edxapp/templates/services/flower/dc.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  template:
+    metadata:
+      labels:
+        app: edxapp
+        service: flower
+        version: "{{ edxapp_flower_image_tag }}"
+        deployment_stamp: "{{ deployment_stamp }}"
+        deploymentconfig: "edxapp-flower-{{ deployment_stamp }}"
+    spec:
+      # Prefer running pods on different nodes for redundancy
```
Comment (madmatah) [inline apps/edxapp/templates/services/flower/dc.yml.j2]: Since we have only 1 replica, is it useful to specify a podAntiAffinity ?

Comment (lunika) [reply apps/edxapp/templates/services/flower/dc.yml.j2]: No this can be removed

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,10 @@ Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
 
+### Added
+
+- flower is available as a service of edxapp application
```
Comment (jmaupetit) [inline CHANGELOG.md]:
```suggestion
- `flower` is available as a service of `edxapp` application
```

File: apps/edxapp/templates/services/flower/configs/project.py.j2 [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from celery import Celery
+from celery_redis_sentinel import register
+
+register()
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/flower/configs/project.py.j2]: I would had a comment explaining why your are monkey patching celery bootstraping.

File: apps/edxapp/templates/services/nginx/configs/flower.conf.j2 [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        $http_x_forwarded_for variable contains a list of IPs listed from the HTTP_X_FORWARED_FOR 
+        header (e.g. w.x.y.z, 10.0.0.1). The first IP corresponds to the client's public address,
+        which is of interest (other ones have been added by subsequent proxies),
+        hence we restrict our comparison with the beginning of this list (this is why our regex starts with a ^).
+    #}
+    if ($http_x_forwarded_for !~ ^({{ edxapp_nginx_flower_ip_whitelist | join("|") }})) {
+      return 403;
+    }
+{% endif %}
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/flower.conf.j2]: So there is no HTTP basic auth, only an IP whitelisting?

File: apps/edxapp/vars/all/main.yml [hunk 5]
```diff
@@ -103,13 +104,22 @@ edxapp_nginx_image_name: "fundocker/openshift-nginx"
...
 edxapp_nginx_healthcheck_endpoint: "/__healthcheck__"
 edxapp_routing_timeout: "60s"
 edxapp_nginx_cms_admin_ip_withelist: []
 edxapp_nginx_lms_admin_ip_withelist: []
+edxapp_nginx_flower_ip_whitelist: []
+
+# -- flower
+
+edxapp_flower_image_name: "fundocker/flower"
+edxapp_flower_image_tag: "1.0"
```
Comment (jmaupetit) [inline apps/edxapp/vars/all/main.yml]: Should be fixed once we released the first version of our image.

**Review by jmaupetit — COMMENTED** [review]
I have the feeling that each flower instance (current, next or previous) will see all workers from all stacks of the project (current, next or previous). Is there a way to restrict data pooling by the workers host name or deployment stamp?

File: apps/edxapp/templates/services/nginx/configs/flower.conf.j2 (hunk 4 above)
Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/flower.conf.j2]: yes

File: apps/flower/templates/services/nginx/configs/flower.conf.j2 [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+        $http_x_forwarded_for variable contains a list of IPs listed from the HTTP_X_FORWARED_FOR 
+        header (e.g. w.x.y.z, 10.0.0.1). The first IP corresponds to the client's public address,
+        which is of interest (other ones have been added by subsequent proxies),
+        hence we restrict our comparison with the beginning of this list (this is why our regex starts with a ^).
+    #}
+    if ($http_x_forwarded_for !~ ^({{ flower_nginx_app_ip_whitelist | join("|") }})) {
+      return 403;
+    }
+{% endif %}
```
Comment (jmaupetit) [inline apps/flower/templates/services/nginx/configs/flower.conf.j2:32]: Not sure we should let it open by default...

File: apps/flower/vars/settings.yml [hunk 7]
```diff
@@ -0,0 +1 @@
+is_blue_green_compatible: False
```
Comment (jmaupetit) [inline apps/flower/vars/settings.yml]:
```suggestion
is_blue_green_compatible: false
```

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**